### DISCUSSION TITLE
Integrated neural ODE or neural state space

### DISCUSSION THREAD

**Discussion by JIEZHUZED** [do-mpc/do-mpc#380] [Ideas]
Hi!

I noticed that `ONNXConversion` was added in release 4.5.0 can convert a previously trained and stored neural network to a `CasADi `graph.

If I want to integrate a `Torchdyn` or `Matlab` trained neural ODE model with do-mpc is this feasible?

For example, I try to output the trained model as ONNX, or MLP re-implemented using Casadi, and later build the model using `set_rhs()`.

I am still a beginner not sure if this idea is feasible. It would be great if someone could help me.

Thanks

**Comment by 4flixt** [comment]
Hi,

@JIEZHUZED, this is indeed already possible - although the feature is still experimental and has some limitations. We will soon upload an example. 
The process then involves the following steps:
- export your trained neural network via ONNX
- use the ONNXConversion tool to obtain an equivalent expression in the CasADi notation
- use this formulation anywhere in your model

Best,
@4flixt

**Reply by JIEZHUZED** [reply]
Thank you so much, looking forward to this example.

**Reply by 4flixt** [reply]
Hi @JIEZHUZED,

we have just published do-mpc v4.6.0 with the official inclusion of the ONNXConversion tools. You can learn more about them in our [documentation](https://www.do-mpc.com/en/latest/api/do_mpc.sysid.ONNXConversion.html#do_mpc.sysid.ONNXConversion).

Following this link, you will also find a brief example of how  to use them. We are currently working on a more complete Jupyter Notebook tutorial, but depending on your prior knowledge the existing docs might be sufficient to get you started.

Best,
4flixt
